design-proposals/defaulting.md

```diff
@@ -0,0 +1,122 @@
+```
+
+After we have defined the domain in libvirt, the XML looks like this:
+
+```xml
+<domain type='qemu' xmlns:qemu='http://libvirt.org/schemas/domain/qemu/1.0'>
+  <name>testvm</name>
+  <uuid>a8335812-53d4-4e57-ac0b-72951c885a57</uuid>
+  <memory unit='KiB'>62500</memory>
+  <currentMemory unit='KiB'>62500</currentMemory>
+  <vcpu placement='static'>1</vcpu>
+  <os>
+    <type arch='x86_64' machine='pc-i440fx-2.7'>hvm</type>
+    <boot dev='hd'/>
+  </os>
+  <clock offset='utc'/>
+  <on_poweroff>destroy</on_poweroff>
+  <on_reboot>restart</on_reboot>
+  <on_crash>destroy</on_crash>
+  <devices>
+    <emulator>/usr/bin/qemu-system-x86_64</emulator>
+    <controller type='usb' index='0' model='piix3-uhci'>
+      <address type='pci' domain='0x0000' bus='0x00' slot='0x01' function='0x2'/>
+    </controller>
+    <controller type='pci' index='0' model='pci-root'/>
+    <interface type='network'>
+      <mac address='52:54:00:2e:84:30'/>
+      <source network='default'/>
+      <model type='rtl8139'/>
+      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0'/>
+    </interface>
+    <input type='mouse' bus='ps2'/>
+    <input type='keyboard' bus='ps2'/>
+    <memballoon model='virtio'>
+      <address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
+    </memballoon>
+  </devices>
+</domain>
+```
+
+Finally, when a Domain was successfully started by Libvirt, Libvirt assigns an `alias` to every device in the XML:
+
+## The Problem
+
+ * Hotplug operations are not possible with the minimal VM spec of kubevirt, since we could match wrong devices.
+ * After VM restarts, different devices or addresses could be assigned to a VM by libvirt
+ * Libvirt assigns all that data on the host, either when defining the Domain, or after the Domain was started.
+   That requires asynchronous lookups, and when thinking about administrating offline VMs, it binds your domain specification management to the kubevirt runtime. happen, to get that data
+
+If your VMs are ephemeral (no migrations, recreating VMs vs. restarting VMs, no hotplug), you are not affected by that.
+
+## Proposed Solution
+
+### Defaulting service based on libvirt
+
+Except for some runtime information, which is normally not included in migratable domain XMLs, libvirt fills in all defaults when defining a domain. We can leverage that with a simple REST based service ,which defines VMs in an isolated libvirt (running libvirt and the small REST service in a Pod). The service is a standalone component in KubeVirt and can then be integrated into two different flows:
+
+ 1. When doing a POST to `/apis/kubevirt.io/v1alpha1/namespaces/mynamespace/vms`, `virt-api` can do a roundtrip to that service and prefill the defaults. As the response of the POST to `virt-api`, you will get the completely prefilled VM specification. This updated spec contains all the necessary bits, to allow consistent hotplugging and keeping the VM definition consistent between redefines.
```

Comment:
@berrange I agree that in many cases you exactly want the behaviour you described. I would also actually favour making this defaulting opt-in. 

But the oVirt VMs are the perfect example that there is the need for keeping the devices like they were generated the first time. Basically the VM also sees adresses, ... and changes to that settings can be disruptive for the OS in the VM.
Maybe we can narrow down the scope of the defaulter a little bit, and we could find parts which are defaulted but not visible from inside the VM. Then we can leave them out.

Also, to avoid misunderstandings, the described POST URL in 1 is only about VMs in the cluster which should run. This is not the place to store VMs for later runs (you can think of it a little bit like transient VMs in libvirt, except that they do not immediately disappear if the crash). So I completely agree to

> those defaults should not be saved back into the master config stored in K8s API, but to a separate area, so if you stop & start the VM again later, it picks up the newer defaults once more.

If you want to make use of these default for later runs is completely optional. The goal I had in mind is to allow VM stores on top of this KubeVirt runtime environment, to choose which level of consistency they want to provide to the outside world, without having to asynchronously collect defaults assigned by the lowest layers in the stack.